File: lib/_union_find.py
class UnionFind:
    def __init__(self):
        self.parents = {}

    def union(self, x, y):
        self.parents[self.find(x)] = self.find(y)

    def find(self, x):
        if x not in self.parents or self.parents[x] == x:
            return x
        self.parents[x] = self.find(self.parents[x])
        return self.parents[x]

File: src/lib/unionfind.ts
import { ValueMap } from "./collections";

export class UnionFind<T> {
    private parents: ValueMap<T, T>;

    constructor() {
        this.parents = new ValueMap<T, T>([]);
    }

    union(x: T, y: T) {
        this.parents.set(this.find(x), this.find(y));
    }

    find(x: T): T {
        if (!this.parents.has(x) || this.parents.get(x) === x) {
            return x;
        }
        this.parents.set(x, this.find(this.parents.get(x)));
        return this.parents.get(x);
    }
}
